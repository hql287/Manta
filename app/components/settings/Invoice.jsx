// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
const ipc = require('electron').ipcRenderer;
import { keys, sortBy } from 'lodash';
const moment = require('moment');

// Custom Libs
import currencies from '../../../libs/currencies.json';
const openDialog = require('../../renderers/dialog.js');
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

import styled from 'styled-components';

const Row = styled.div`
  display: flex;
  margin: 0 -15px;
`;

const Field = styled.div`
  flex: 1;
  margin: 0 15px 20px 15px;
`;

const Header = styled.h2``;

const Section = styled.div`
  padding: 20px;
  background: #f9fafa;
  border-radius: 4px;
  margin-bottom: 20px;
  border: 1px solid #f2f3f4;
`;

// Component
class Invoice extends Component {
  constructor(props) {
    super(props);
    this.state = this.props.invoice;
    this.selectExportDir = this.selectExportDir.bind(this);
    this.sortCurrencies = this.sortCurrencies.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleTaxChange = this.handleTaxChange.bind(this);
    this.handleVisibilityChange = this.handleVisibilityChange.bind(this);
    this.t = this.props.t;
  }

  componentDidMount() {
    ipc.on('no-access-directory', (event, message) => {
      openDialog({
        type: 'warning',
        title: this.t('settings:invoice:noaccess:title'),
        message: message + this.t('settings:invoice:noaccess:message'),
      });
    });

    ipc.on('confirmed-export-directory', (event, path) => {
      this.setState({ exportDir: path }, () => {
        this.props.updateSettings('invoice', this.state);
      });
    });
  }

  componentWillUnmount() {
    ipc.removeAllListeners('no-access-directory');
    ipc.removeAllListeners('confirmed-export-directory');
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState({ [name]: value }, () => {
      this.props.updateSettings('invoice', this.state);
    });
  }

  handleTaxChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'amount' ? parseFloat(target.value) : target.value;
    this.setState(
      {
        tax: Object.assign({}, this.state.tax, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  handleVisibilityChange(event) {
    const target = event.target;
    const name = target.name;
    const value = target.checked;
    this.setState(
      {
        required_fields: Object.assign({}, this.state.required_fields, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  selectExportDir() {
    ipc.send('select-export-directory');
  }

  sortCurrencies() {
    const currenciesKeys = keys(currencies);
    const currenciesKeysAndValues = currenciesKeys.map(key => [
      key,
      currencies[key].name,
      currencies[key].code,
    ]);
    const currenciesSorted = sortBy(currenciesKeysAndValues, [
      array => array[1],
    ]);
    return currenciesSorted.map(obj => {
      const [key, name, code] = obj;

      const optionKey = code;
      const optionValue = code;
      const optionLabel = name;

      return (
        <option value={optionValue} key={optionKey}>
          {optionLabel}
        </option>
      );
    });
  }

  render() {
    const {
      exportDir,
      template,
      currency,
      tax,
      required_fields,
      dateFormat,
    } = this.state;
    return (
      <div>
        <label className="itemLabel">{this.t('settings:invoice:tax:heading')}</label>
        <Section>
          <Row>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:tax:taxid')}</label>
              <input
                name="tin"
                type="text"
                value={tax.tin}
                onChange={this.handleTaxChange}
                placeholder={this.t('settings:invoice:tax:taxidplaceholder')}
              />
            </Field>
          </Row>
          <Row>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:tax:amount')}</label>
              <input
                name="amount"
                type="number"
                step="0.01"
                value={tax.amount}
                onChange={this.handleTaxChange}
                placeholder="Amount"
              />
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:tax:method:heading')}</label>
              <select
                name="method"
                value={tax.method}
                onChange={this.handleTaxChange}
              >
                <option value="default">{this.t('settings:invoice:tax:method:default')}</option>
                <option value="reverse">{this.t('settings:invoice:tax:method:reverse')}</option>
              </select>
            </Field>
          </Row>
        </Section>

        <label className="itemLabel">{this.t('settings:invoice:reqfields:heading')}</label>
        <Section>
          <Row>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:reqfields:duedate')}</label>
              <label className="switch">
                <input
                  name="dueDate"
                  type="checkbox"
                  checked={required_fields.dueDate}
                  onChange={this.handleVisibilityChange}
                />
                <span className="slider round" />
              </label>
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:common:currency')}</label>
              <label className="switch">
                <input
                  name="currency"
                  type="checkbox"
                  checked={required_fields.currency}
                  onChange={this.handleVisibilityChange}
                />
                <span className="slider round" />
              </label>
            </Field>

            <Field>
              <label className="itemLabel">{this.t('settings:invoice:reqfields:discount')}</label>
              <label className="switch">
                <input
                  name="discount"
                  type="checkbox"
                  checked={required_fields.discount}
                  onChange={this.handleVisibilityChange}
                />
                <span className="slider round" />
              </label>
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:reqfields:tax')}</label>
              <label className="switch">
                <input
                  name="tax"
                  type="checkbox"
                  checked={required_fields.tax}
                  onChange={this.handleVisibilityChange}
                />
                <span className="slider round" />
              </label>
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:reqfields:note')}</label>
              <label className="switch">
                <input
                  name="note"
                  type="checkbox"
                  checked={required_fields.note}
                  onChange={this.handleVisibilityChange}
                />
                <span className="slider round" />
              </label>
            </Field>
          </Row>
        </Section>

        <label className="itemLabel">{this.t('settings:invoice:other:heading')}</label>
        <Section>
          <Row>
            <Field>
              <label className="itemLabel">{this.t('settings:common:currency')}</label>
              <select
                name="currency"
                value={currency}
                onChange={this.handleInputChange}
              >
                {this.sortCurrencies()}
              </select>
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:other:template:heading')}</label>
              <select
                name="template"
                value={template}
                onChange={this.handleInputChange}
              >
                <option value="minimal">{this.t('settings:invoice:other:template:minimal')}</option>
                <option value="business">{this.t('settings:invoice:other:template:business')}</option>
              </select>
            </Field>
          </Row>
          <Row>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:other:dateformat')}</label>
              <select
                name="dateFormat"
                value={dateFormat}
                onChange={this.handleInputChange}
              >
                <option value="dddd, MMMM Do, YYYY">
                  {moment(Date.now()).format('dddd, MMMM Do, YYYY')} (dddd, MMMM
                  Do, YYYY)
                </option>
                <option value="MMMM Do, YYYY">
                  {moment(Date.now()).format('MMMM Do, YYYY')} (MMMM Do, YYYY)
                </option>
                <option value="MM/DD/YYYY">
                  {moment(Date.now()).format('MM/DD/YYYY')} (MM/DD/YYYY)
                </option>
                <option value="MM/DD/YY">
                  {moment(Date.now()).format('MM/DD/YY')} (MM/DD/YY)
                </option>
                <option value="dddd, DD MMMM YYYY">
                  {moment(Date.now()).format('dddd, DD MMMM YYYY')} (dddd, DD
                  MMMM YYYY)
                </option>
                <option value="DD/MMMM/YYYY">
                  {moment(Date.now()).format('DD/MMMM/YYYY')} (DD/MMMM/YYYY)
                </option>
                <option value="DD/MM/YYYY">
                  {moment(Date.now()).format('DD/MM/YYYY')} (DD/MM/YYYY)
                </option>
                <option value="DD/MM/YY">
                  {moment(Date.now()).format('DD/MM/YY')} (DD/MM/YY)
                </option>
              </select>
            </Field>
            <Field>
              <label className="itemLabel">{this.t('settings:invoice:other:exportdir')}</label>
              <div className="input-group">
                <input
                  className="form-control"
                  name="exportDir"
                  type="text"
                  value={exportDir}
                  disabled
                />
                <a
                  href="#"
                  className="input-group-customized "
                  onClick={this.selectExportDir}
                >
                  <i className="ion-folder" />
                </a>
              </div>
            </Field>
          </Row>
        </Section>
      </div>
    );
  }
}

Invoice.propTypes = {
  invoice: PropTypes.object.isRequired,
  updateSettings: PropTypes.func.isRequired,
};

export default _withFadeInAnimation(Invoice);
