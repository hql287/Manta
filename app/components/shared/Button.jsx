// Libraries
import React from 'react';
import PropTypes from 'prop-types';

// Styles
import styled from 'styled-components';

const ButtonStyle = styled.button`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  padding: 4px 15px;
  font-size: 12px;
  text-decoration: none;
  background: #ffffff;
  border: 1px solid rgba(0,0,0,.1);
  text-transform: uppercase;
  letter-spacing: 1px;
  ${props => props.block && `width: 100%;`}
  ${props => props.primary && `
    background: #469fe5;
    color: white;
  `}
  ${props => props.success && `
    background: #6bbb69;
    color: white;
  `}
  ${props => props.danger && `
    background: #EC476E;
    color: white;
  `}
  &:hover {
    cursor: pointer;
    // color: white;
    text-decoration: none;
  }
`;

const ButtonLinkStyle = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  ${props => props.primary && `color: #469fe5;`}
  ${props => props.success && `color: #6bbb69;`}
  ${props => props.danger && `color: #EC476E;`}
  &:hover { cursor: pointer; }
`;

function Button(props) {
  return props.link
    ? <ButtonLinkStyle {...props}>
        {props.children}
      </ButtonLinkStyle>
    : <ButtonStyle {...props}>
        {props.children}
      </ButtonStyle>;
}

Button.propTypes = {
  link: PropTypes.bool,
  danger: PropTypes.bool,
  onClick: PropTypes.func,
  primary: PropTypes.bool,
  success: PropTypes.bool,
};

Button.defaultProps = {
  link: false,
  primary: false,
  success: false,
  danger: false,
};

export default Button;
