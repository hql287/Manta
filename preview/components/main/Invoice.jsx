// Libraries
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

// Style
import styled from 'styled-components';

const Page = styled.div`
  position: relative;
  width: 21cm;
  height: 29.7cm;
  min-height: 29.7cm;
  min-width: 21cm;
  margin-left: auto;
  margin-right: auto;
  background: #FFFFFF;
  box-shadow: 0 0 10px rgba(0,0,0,.1);
  display: flex;
  border-radius: 4px;
}
`;

// Templates
import Minimal from '../../templates/minimal';
import Business from '../../templates/business';

// Component
class Invoice extends PureComponent {

  renderTemplate() {
    switch (this.props.configs.template) {
      case 'business': {
        return <Business {...this.props} />;
      }
      default: {
        return <Minimal {...this.props} />;
      }
    }
  }

  render() {
    return <Page>{this.renderTemplate()}</Page>;
  }
}

Invoice.propTypes = {
  configs: PropTypes.object.isRequired,
  invoice: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
};

export default Invoice;
