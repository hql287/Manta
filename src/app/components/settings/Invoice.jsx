// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
const ipc = require('electron').ipcRenderer;

// Custom Libs
const openDialog = require('../../renderers/dialog.js');
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

import Currency from './_partials/invoice/Currency';
import Fields from './_partials/invoice/Fields';
import Other from './_partials/invoice/Other';
import Tax from './_partials/invoice/Tax';

// Component
class Invoice extends Component {
  constructor(props) {
    super(props);
    this.state = this.props.invoice;
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleTaxChange = this.handleTaxChange.bind(this);
    this.handleCurrencyChange = this.handleCurrencyChange.bind(this);
    this.handleVisibilityChange = this.handleVisibilityChange.bind(this);
  }

  componentDidMount() {
    const { t } = this.props;
    ipc.on('no-access-directory', (event, message) => {
      openDialog({
        type: 'warning',
        title: t('dialog:noAccess:title'),
        message: `${message}. ${t('dialog:noAccess:message')}`,
      });
    });

    ipc.on('confirmed-export-directory', (event, path) => {
      this.setState({ exportDir: path }, () => {
        this.props.updateSettings('invoice', this.state);
      });
    });
  }

  componentWillUnmount() {
    ipc.removeAllListeners('no-access-directory');
    ipc.removeAllListeners('confirmed-export-directory');
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState({ [name]: value }, () => {
      this.props.updateSettings('invoice', this.state);
    });
  }

  handleTaxChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'amount' ? parseFloat(target.value) : target.value;
    this.setState(
      {
        tax: Object.assign({}, this.state.tax, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  handleCurrencyChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'fraction' ? parseInt(target.value, 10) : target.value;
    this.setState(
      {
        currency: Object.assign({}, this.state.currency, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  handleVisibilityChange(event) {
    const target = event.target;
    const name = target.name;
    const value = target.checked;
    this.setState(
      {
        required_fields: Object.assign({}, this.state.required_fields, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  selectExportDir() {
    ipc.send('select-export-directory');
  }

  render() {
    const { t } = this.props;
    const {
      exportDir,
      template,
      currency,
      tax,
      required_fields,
      dateFormat,
    } = this.state;
    return [
      <Fields
        key="required_fields_settings"
        required_fields={required_fields}
        handleVisibilityChange={this.handleVisibilityChange}
        t={t}
      />,
      <Tax
        key="tax_settings"
        handleTaxChange={this.handleTaxChange}
        tax={tax}
        t={t}
      />,
      <Currency
        key="currency_settings"
        currency={currency}
        handleCurrencyChange={this.handleCurrencyChange}
        t={t}
      />,
      <Other
        key="other_settings"
        dateFormat={dateFormat}
        exportDir={exportDir}
        template={template}
        handleInputChange={this.handleInputChange}
        selectExportDir={this.selectExportDir}
        t={t}
      />
    ];
  }
}

Invoice.propTypes = {
  invoice: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateSettings: PropTypes.func.isRequired,
};

export default _withFadeInAnimation(Invoice);
