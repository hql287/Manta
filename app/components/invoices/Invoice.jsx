// Libs
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { truncate } from 'lodash';
import styled from 'styled-components';
const moment = require('moment');
const ipc = require('electron').ipcRenderer;

// Helper
import { formatNumber } from '../../../helpers/formatNumber';
import { calTermDate } from '../../../helpers/date';

// Custom Components
import Button from '../shared/Button';
import SplitButton from '../shared/SplitButton';

// Invoice Container
const Wrapper = styled.div`
  position: relative;
  border: 1px solid rgba(0, 0, 0, 0.1);
  background: white;
  margin-bottom: 30px;
  border-radius: 4px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
`;

// Invoice Header
const Header = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 25px 30px;
`;

const StatusBar = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 6px;
  border-radius: 4px 4px 0 0;
  ${props => props.status === 'pending' && `background: #469FE5;`} ${props =>
      props.status === 'paid' && `background: #6BBB69;`} ${props =>
      props.status === 'refunded' && `background: #4F555C;`} ${props =>
      props.status === 'cancelled' && `background: #EC476E;`};
`;

const Status = styled.div`
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  letter-spacing: 1px;
  ${props => props.status === 'pending' && `color: #469FE5;`} ${props =>
      props.status === 'paid' && `color: #6BBB69;`} ${props =>
      props.status === 'refunded' && `color: #4F555C;`} ${props =>
      props.status === 'cancelled' && `color: #EC476E;`} span {
    display: flex;
    align-items: center;
    i {
      margin-right: 5px;
    }
  }
  i.ion-checkmark {
    font-size: 16px;
    line-height: 16px;
  }
  i.ion-loop {
    font-size: 18px;
    line-height: 18px;
  }
  i.ion-backspace {
    font-size: 18px;
    line-height: 18px;
  }
  i.ion-arrow-return-left {
    font-size: 18px;
    line-height: 18px;
  }
`;

const ButtonsGroup = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;
  i {
    margin-left: 10px;
    color: #B4B7BA;
  }
  i.ion-trash-a {
    font-size: 24px;
    line-height: 24px;
    &:hover {
      color: #EC476E;
    }
  }
  i.ion-ios-copy {
    font-size: 24px;
    line-height: 24px;
    &:hover {
      color: #469FE5;
    }
  }
}
`;

// Invoice Body
const Body = styled.div`
  padding: 0 30px;
`;

// Invoice Footer
const Footer = styled.div`
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  > * {
    flex: 1;
    margin: 0 10px;
    &:first-child {
      flex: 3;
    }
  }
`;

// Shared Style
const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
`;

const Field = styled.div`
  position: relative;
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  margin-bottom: 15px;
  text-transform: capitalize;
  h2 {
    font-size: 21px;
    color: #283641;
    margin-bottom: 0;
    font-weight: 500;
  }
  label {
    font-size: 11px;
    color: #b4b7ba;
    text-transform: uppercase;
    letter-spacing: 1px;
    margin-bottom: 4px;
    font-weight: 400;
  }
  p {
    font-weight: 300;
    margin-bottom: 0px;
    font-size: 14px;
  }
`;

// Component
class Invoice extends PureComponent {
  constructor(props) {
    super(props);
    this.viewInvoice = this.viewInvoice.bind(this);
    this.editInvoice = this.editInvoice.bind(this);
    this.deleteInvoice = this.deleteInvoice.bind(this);
    this.duplicateInvoice = this.duplicateInvoice.bind(this);
    this.displayStatus = this.displayStatus.bind(this);
  }

  deleteInvoice() {
    const { invoice, deleteInvoice } = this.props;
    deleteInvoice(invoice._id);
  }

  duplicateInvoice() {
    const { invoice, duplicateInvoice } = this.props;
    duplicateInvoice(invoice);
  }

  editInvoice() {
    const { invoice, editInvoice } = this.props;
    editInvoice(invoice);
  }

  viewInvoice() {
    ipc.send('preview-invoice', this.props.invoice);
  }

  displayStatus() {
    const { t, invoice } = this.props;
    const { status } = invoice;
    const { recipient } = invoice;
    switch (status) {
      case 'cancelled': {
        return (
          <span>
            <i className="ion-backspace" />
            {t('invoices:status:cancelled')}
          </span>
        );
      }
      case 'paid': {
        return (
          <span>
            <i className="ion-checkmark" />
            {t('invoices:status:paid')}
          </span>
        );
      }
      case 'refunded': {
        return (
          <span>
            <i className="ion-arrow-return-left" />
            {t('invoices:status:refunded')}
          </span>
        );
      }
      default: {
        return (
          <span>
            <i className="ion-loop" />
            {t('invoices:status:pending')}
          </span>
        );
      }
    }
  }

  renderDueDate() {
    const { t, invoice } = this.props;
    const { dueDate, configs } = invoice;
    const { useCustom, paymentTerm, selectedDate } = dueDate;
    const dateFormat = configs ? configs.dateFormat : this.props.dateFormat;
    // If it's a custom date then return selectedDate
    if (useCustom === true) {
      return  moment(selectedDate).format(dateFormat);
    }
    // If it's a payment term, calculate the term date and print out
    const paymentTermDate = calTermDate(invoice.created_at, paymentTerm);
    return `
      ${ t(`form:fields:dueDate:paymentTerms:${paymentTerm}:label`) }
      (
      ${ moment(paymentTermDate).format(dateFormat) }
      )
    `;
  }

  render() {
    const { invoice, setInvoiceStatus, t } = this.props;
    const { recipient, status, configs } = invoice;
    const dateFormat = invoice.configs ? invoice.configs.dateFormat : this.props.dateFormat;
    const statusActions = [
      {
        label: t('invoices:status:pending'),
        action: () => setInvoiceStatus(invoice._id, 'pending'),
      },
      {
        label: t('invoices:status:refunded'),
        action: () => setInvoiceStatus(invoice._id, 'refunded'),
      },
      {
        label: t('invoices:status:cancelled'),
        action: () => setInvoiceStatus(invoice._id, 'cancelled'),
      },
    ];
    const currencyBefore = invoice.currency.placement === 'before';
    return (
      <div className="col-lg-6">
        <Wrapper>
          <StatusBar status={status} />
          <Header>
            <Status status={status}>{this.displayStatus()}</Status>
            <ButtonsGroup>
              <Button link onClick={this.duplicateInvoice}>
                <i className="ion-ios-copy" />
              </Button>
              <Button link onClick={this.deleteInvoice}>
                <i className="ion-trash-a" />
              </Button>
            </ButtonsGroup>
          </Header>
          <Body>
            <Row>
              <Field>
                <label>{t('invoices:fields:client')}</label>
                <h2>{recipient.fullname}</h2>
              </Field>
            </Row>
            <Row>
              <Field>
                <label>{t('invoices:fields:invoiceID')}</label>
                <p>
                  {invoice.invoiceID
                    ? invoice.invoiceID
                    : truncate(invoice._id, {
                        length: 8,
                        omission: '',
                      })}
                </p>
              </Field>
              <Field>
                <label>{t('invoices:fields:total')}</label>
                <p>
                  {currencyBefore ? invoice.currency.code : null}
                  {' '}
                  { formatNumber(
                      invoice.grandTotal,
                      invoice.currency.fraction,
                      invoice.currency.separator)
                  }
                  {' '}
                  {currencyBefore ? null : invoice.currency.code}
                </p>
              </Field>
            </Row>
            <Row>
              <Field>
                <label>{t('invoices:fields:createdDate')}</label>
                <p>{moment(invoice.created_at).format(dateFormat)}</p>
              </Field>
              <Field>
                <label>{t('invoices:fields:dueDate')}</label>
                <p>
                  { invoice.dueDate && this.renderDueDate() }
                </p>
              </Field>
            </Row>
          </Body>
          <Footer>
            <SplitButton
              mainButton={{
                label: t('invoices:btns:markAsPaid'),
                action: () => setInvoiceStatus(invoice._id, 'paid'),
              }}
              options={statusActions}
            />
            <Button onClick={this.editInvoice}>
              {t('invoices:btns:edit')}
            </Button>
            <Button onClick={this.viewInvoice}>
              {t('invoices:btns:view')}
            </Button>
          </Footer>
        </Wrapper>
      </div>
    );
  }
}

Invoice.propTypes = {
  dateFormat: PropTypes.string.isRequired,
  deleteInvoice: PropTypes.func.isRequired,
  duplicateInvoice: PropTypes.func.isRequired,
  editInvoice: PropTypes.func.isRequired,
  invoice: PropTypes.object.isRequired,
  setInvoiceStatus: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
};

export default Invoice;
