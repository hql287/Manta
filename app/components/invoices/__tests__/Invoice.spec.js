// Libs
import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';

// Component
import Invoice from '../Invoice';
import Button from '../../shared/Button';
import SplitButton from '../../shared/SplitButton';

// Mocks
const invoice = {
  _id: '37b2804e-bfc1-4289-b1d7-226c5652ac91',
  status: 'pending',
  currency: {
    code: 'USD',
    decimal_digits: 2,
    name: 'US Dollar',
    name_plural: 'US dollars',
    rounding: 0,
    symbol: '$',
    symbol_native: '$',
  },
  subtotal: 3843,
  grandTotal: 3843,
  recipient: {
    _id: '3bd85cb9-7675-4d59-b9c3-305481cb77c9',
    company: 'Lindgren Group',
    email: 'Hattie42@gmail.com',
    fullname: 'Penelope Hettinger',
    phone: '854-306-7837',
  },
};

const deleteInvoice = jest.fn();
const setInvoiceStatus = jest.fn();

// Tests
describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Invoice
        invoice={invoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
  });

  it('receives correct props', () => {
    const mountWrapper = mount(
      <Invoice
        invoice={invoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
    expect(mountWrapper.prop('invoice')).toEqual(invoice);
    expect(mountWrapper.prop('deleteInvoice')).toEqual(deleteInvoice);
    expect(mountWrapper.prop('setInvoiceStatus')).toEqual(setInvoiceStatus);
  });

  it('render with a class of .col-lg-6', () => {
    expect(wrapper.find('.col-lg-6')).toHaveLength(1);
    expect(wrapper.find('.col-lg-6')).not.toHaveLength(2);
  });

  it('render invoice status in the header', () => {
    expect(
      wrapper.find('Invoice__Header').find('Invoice__Status'),
    ).toHaveLength(1);
  });

  it('render invoice status correctly', () => {
    // Pending Invoice (default)
    expect(
      wrapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text(),
    ).toEqual('Pending');

    // Paid Invoice
    const paidInvoice = Object.assign({}, invoice, {status: 'paid'});
    const paidInvoiceWapper = shallow(
      <Invoice
        invoice={paidInvoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
    expect(
      paidInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text(),
    ).toEqual('Paid');

    // Cancelled Invoice
    const cancelledInvoice = Object.assign({}, invoice, {status: 'cancelled'});
    const cancelledInvoiceWapper = shallow(
      <Invoice
        invoice={cancelledInvoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
    expect(
      cancelledInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text(),
    ).toEqual('Cancelled');

    // Refunded Invoice
    const refundedInvoice = Object.assign({}, invoice, {status: 'refunded'});
    const refundedInvoiceWapper = shallow(
      <Invoice
        invoice={refundedInvoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
    expect(
      refundedInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text(),
    ).toEqual('Refunded');
  });

  it('render delete button in the header', () => {
    expect(wrapper.find('Invoice__Header').find(Button)).toHaveLength(1);
  });

  it('render a split button and 2 buttons in the footer', () => {
    expect(wrapper.find('Invoice__Footer').find(SplitButton)).toHaveLength(1);
    expect(wrapper.find('Invoice__Footer').find(Button)).toHaveLength(2);
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Invoice
          invoice={invoice}
          deleteInvoice={deleteInvoice}
          setInvoiceStatus={setInvoiceStatus}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('handle clicks correctly', () => {
  let wrapper, editBtn, viewBtn, deleteBtn, spyViewAction, spyEditAction;
  beforeEach(() => {
    spyViewAction = jest.spyOn(Invoice.prototype, 'viewInvoice');
    spyEditAction = jest.spyOn(Invoice.prototype, 'editInvoice');
    wrapper = shallow(
      <Invoice
        invoice={invoice}
        deleteInvoice={deleteInvoice}
        setInvoiceStatus={setInvoiceStatus}
      />,
    );
    viewBtn = wrapper
      .find('Invoice__Footer')
      .find(Button)
      .last();
    editBtn = wrapper
      .find('Invoice__Footer')
      .find(Button)
      .first();
    deleteBtn = wrapper.find(Button).first();
  });

  it('handle delete action correctly', () => {
    deleteBtn.simulate('click');
    expect(deleteInvoice).toHaveBeenCalled();
    expect(deleteInvoice).toHaveBeenCalledWith(invoice._id);
  });

  it('handle view action correctly', () => {
    viewBtn.simulate('click');
    expect(spyViewAction).toHaveBeenCalled();
  });

  it('handle edit action correctly', () => {
    editBtn.simulate('click');
    expect(spyEditAction).toHaveBeenCalled();
  });
});
