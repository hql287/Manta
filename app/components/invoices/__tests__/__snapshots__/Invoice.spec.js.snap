// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="col-lg-6"
>
  <div
    className="Invoice__Wrapper-iBPCPb iuhTYc"
  >
    <div
      className="Invoice__StatusBar-gmZFuj kFdsGo"
    />
    <div
      className="Invoice__Header-cAhbPy ePNeRZ"
    >
      <div
        className="Invoice__Status-eKPgUx eiULaG"
      >
        <span>
          <i
            className="ion-loop"
          />
          Pending
        </span>
      </div>
      <button
        className="Button__ButtonLinkStyle-ioTbaT eFjcaV"
        onClick={[Function]}
      >
        <i
          className="ion-trash-a"
        />
      </button>
    </div>
    <div
      className="Invoice__Body-dSsrWx qPyCt"
    >
      <div
        className="Invoice__Row-gNlmug hPTbEG"
      >
        <div
          className="Invoice__Field-iXxbne toUew"
        >
          <label>
            Client
          </label>
          <h2>
            Penelope Hettinger
          </h2>
        </div>
      </div>
      <div
        className="Invoice__Row-gNlmug hPTbEG"
      >
        <div
          className="Invoice__Field-iXxbne toUew"
        >
          <label>
            Invoice Id
          </label>
          <p>
            37b2804e
          </p>
        </div>
        <div
          className="Invoice__Field-iXxbne toUew"
        >
          <label>
            Total Value
          </label>
          <p>
            USD
             
            3,843
          </p>
        </div>
      </div>
      <div
        className="Invoice__Row-gNlmug hPTbEG"
      >
        <div
          className="Invoice__Field-iXxbne toUew"
        >
          <label>
            Created On
          </label>
          <p>
            Invalid Date
          </p>
        </div>
        <div
          className="Invoice__Field-iXxbne toUew"
        >
          <label>
            Due Date
          </label>
          <p>
            --
          </p>
        </div>
      </div>
    </div>
    <div
      className="Invoice__Footer-imOzTt jHuxPK"
    >
      <div>
        <div
          className="SplitButton__Wrapper-dWaGGO iZVhhz"
        >
          <div
            className="SplitButton__ButtonsGroup-gPFJVZ jhIoxD"
          >
            <a
              href="#"
              onClick={[Function]}
            >
              Mark As Paid
            </a>
            <a
              href="#"
              onClick={[Function]}
            >
              <i
                className="ion-arrow-down-b"
              />
            </a>
          </div>
        </div>
      </div>
      <button
        className="Button__ButtonStyle-cORGTZ fhgyv"
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="Button__ButtonStyle-cORGTZ fhgyv"
        onClick={[Function]}
      >
        View
      </button>
    </div>
  </div>
</div>
`;
